newrelic/hooks/framework_azurefunctions.py

```diff
@@ -0,0 +1,254 @@
+    faas_name = f"{azure_function_app_name}/{context.function_name}"
+
+    return {
+        "cloud.resource_id": cloud_resource_id,
+        "faas.name": faas_name,
+        "faas.trigger": trigger_type,
+        "faas.invocation_id": context.invocation_id,
+    }
+
+
+# TODO: This should serve as a way to determine the trigger type.
+# Right now, we only support HTTP, so this function serves to activate
+# the application if not already registered with the collector as well
+# as determining if this invocation was a cold start or not.
+async def wrap_dispatcher__handle__invocation_request(wrapped, instance, args, kwargs):
+    # Logic to determine if this is a cold start since we are not
+    # able to access the logic in the __init__ method of the Dispatcher
+    # class with Python (in the Portal, this is not done in Python)
+    if not hasattr(instance, "_nr_running_dispatcher"):
+        instance._nr_running_dispatcher = True
+        instance._nr_cold_start = True
+
+    bound_args = bind_args(wrapped, args, kwargs)
+
+    request = bound_args.get("request", None)
+
+    if not request:
+        return await wrapped(*args, **kwargs)
+
+    # For now, NR only supports HTTP triggers
+    function_id = request.invocation_request.function_id
+
+    binding_type = instance._functions.get_function(function_id).trigger_metadata["type"]
+    if not binding_type.startswith("http"):
+        return await wrapped(*args, **kwargs)
+
+    return await wrapped(*args, **kwargs)
+
+
+async def wrap_dispatcher__run_async_func(wrapped, instance, args, kwargs):
+    from azure.functions.http import HttpRequest
+
+    bound_args = bind_args(wrapped, args, kwargs)
+
+    context = bound_args.get("context", None)
+    func = bound_args.get("func", None)
+    params = bound_args.get("params", None)
+
+    application = original_agent_instance()
+
+    http_request = None
+    for value in params.values():
+        if isinstance(value, HttpRequest):
+            http_request = value
+            url_split = urlparse.urlsplit(getattr(http_request, "url", ":///?#"))
+            scheme = getattr(url_split, "scheme", None)
+            query = getattr(url_split, "query", None)
+            host, port = (
+                (getattr(url_split, "netloc", None), None)
+                if (":" not in getattr(url_split, "netloc", None))
+                else getattr(url_split, "netloc", None).split(":")
+            )
+            break
+
+    # If this is an HTTP http_request object, create a web transaction.
+    # Otherwise, create appropriate type of Transaction in the future
+    if http_request:
+        transaction = WebTransaction(
+            application=application,
+            name=context.function_name,
+            group="AzureFunction",
+            scheme=scheme,
+            host=host,
+            port=port,
+            request_method=http_request.method,
```

Comment:
This should be fine since you're checking the instance type of the `http_request` but just want to ensure safe access of the attributes on it to avoid crashes.